packages/core/src/electron-main/electron-application.ts

```diff
@@ -0,0 +1,440 @@
+     * services global to this process.
+     *
+     * This event is fired when the process starts for the first time.
+     */
+    onStart?(application?: ElectronApplication): MaybePromise<void>;
+    /**
+     * The application is stopping. Contributions must perform only synchronous operations.
+     */
+    onStop?(application?: ElectronApplication): void;
+}
+
+// Extracted the functionality from `yargs@15.4.0-beta.0`.
+// Based on https://github.com/yargs/yargs/blob/522b019c9a50924605986a1e6e0cb716d47bcbca/lib/process-argv.ts
+@injectable()
+export class ProcessArgv {
+
+    protected get processArgvBinIndex(): number {
+        // The binary name is the first command line argument for:
+        // - bundled Electron apps: bin argv1 argv2 ... argvn
+        if (this.isBundledElectronApp) {
+            return 0;
+        }
+        // or the second one (default) for:
+        // - standard node apps: node bin.js argv1 argv2 ... argvn
+        // - unbundled Electron apps: electron bin.js argv1 arg2 ... argvn
+        return 1;
+    }
+
+    protected get isBundledElectronApp(): boolean {
+        // process.defaultApp is either set by electron in an electron unbundled app, or undefined
+        // see https://github.com/electron/electron/blob/master/docs/api/process.md#processdefaultapp-readonly
+        return this.isElectronApp && !(process as ProcessArgv.ElectronProcess).defaultApp;
+    }
+
+    protected get isElectronApp(): boolean {
+        // process.versions.electron is either set by electron, or undefined
+        // see https://github.com/electron/electron/blob/master/docs/api/process.md#processversionselectron-readonly
+        return !!(process as ProcessArgv.ElectronProcess).versions.electron;
+    }
+
+    get processArgvWithoutBin(): Array<string> {
+        return process.argv.slice(this.processArgvBinIndex + 1);
+    }
+
+    get ProcessArgvBin(): string {
+        return process.argv[this.processArgvBinIndex];
+    }
+
+}
+
+export namespace ProcessArgv {
+    export interface ElectronProcess extends NodeJS.Process {
+        readonly defaultApp?: boolean;
+        readonly versions: NodeJS.ProcessVersions & {
+            readonly electron: string;
+        };
+    }
+}
+
+@injectable()
+export class ElectronApplication {
+
+    @inject(ContributionProvider) @named(ElectronMainContribution)
+    protected readonly electronApplicationContributions: ContributionProvider<ElectronMainContribution>;
+
+    @inject(ElectronApplicationGlobals)
+    protected readonly globals: ElectronApplicationGlobals;
+
+    @inject(ElectronSecurityToken)
+    protected electronSecurityToken: ElectronSecurityToken;
+
+    @inject(ProcessArgv)
+    protected processArgv: ProcessArgv;
+
+    protected readonly electronStore = new Storage();
+    protected readonly backendPort = new Deferred<number>();
+
+    protected config: FrontendApplicationConfig;
+
+    async start(config: FrontendApplicationConfig): Promise<void> {
```

Comment:
Could you check in this class which methods should be exposed? I see not many are used outside on the app. What is the reasoning for others?

Reply:
I left `start(config: FrontendApplicationConfig)`, `openNewWindow(url: string)`, `quit()`, and `get config()` as public and reduced the visibility of others.